main/src/library/Graph.flix

```diff
@@ -108,6 +139,49 @@ namespace Graph {
     /// Returns the shortest distance between all pairs of vertices in the
     /// weighted directed graph `g`.
     ///
+    pub def stableDistances(g: m[(t, Int32, t)]): Option[Map[(t, t), Int32]] with Foldable[m], Boxable[t] = {
+        // based on floyd-warshall
+        def rank(v, s) = s |> Set.partition(v0 -> v0 < v) |> fst |> Set.size;
+        let edges = inject g into Edge;
+        let negCycle = #{
+            Node(x) :- Edge(x, _, _).
+            Node(x) :- Edge(_, _, x).
+            // Collect the nodes to find the rank of each node (the number of
+            // nodes less than)
+            Nodes(;Set#{x}) :- Node(x).
+            NodeCount(Set.size(ns)) :- fix Nodes(;ns).
+            NodeRank(x, rank(x, ns)) :- Node(x), fix Nodes(;ns).
+            Dist(x, x, 0; Down(0)) :- Node(x).
+            Dist(x, y, 0; Down(w)) :- Edge(x, w, y).
+            Dist(x, y, k+1; w) :- Dist(x, y, k; w), NodeCount(c), if k < c.
+            Dist(x, y, k+1; w1 + w2) :- Dist(x, kth, k; w1), Dist(kth, y, k; w2),
```

Comment:
I must admit that I simply dont understand this :)

Reply:
I wasn't too generous with the comments yeah. `Dist` is the `ShortestPath` from [wiki](https://en.wikipedia.org/wiki/Floyd%E2%80%93Warshall_algorithm) but I can take a pass with more explanation in comments